flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorSubmissionTest.java

```diff
@@ -555,138 +555,88 @@ public void testFailingScheduleOrUpdateConsumers() throws Exception {
+					tmGateway.requestTaskBackPressure(executionAttemptID, i, timeout);
 
-			assertEquals(response.getSampleId(), sampleId2);
-			assertEquals(response.getExecutionAttemptID(), eid);
+				TaskBackPressureResponse response = successfulRequestFuture.get();
 
-			List<StackTraceElement[]> traces = response.getSamples();
+				assertEquals(response.getRequestId(), i);
+				assertEquals(response.getExecutionAttemptID(), executionAttemptID);
 
-			assertEquals("Number of samples", numSamples, traces.size());
-
-			for (StackTraceElement[] trace : traces) {
-				boolean success = false;
-				for (StackTraceElement elem : trace) {
-					// Look for BlockingNoOpInvokable#invoke
-					if (elem.getClassName().equals(
-						BlockingNoOpInvokable.class.getName())) {
-
-						assertEquals("invoke", elem.getMethodName());
-
-						success = true;
-						break;
-					}
-					// The BlockingNoOpInvokable might not be invoked here
-					if (elem.getClassName().equals(TestTaskManagerActions.class.getName())) {
-
-						assertEquals("updateTaskExecutionState", elem.getMethodName());
-
-						success = true;
-						break;
-					}
-					if (elem.getClassName().equals(Thread.class) && elem.getMethodName().equals("setContextClassLoader")) {
-						success = true;
-					}
+				if ((backPressureRatio = response.getBackPressureRatio()) >= 1.0) {
+					break;
 				}
-
-				assertTrue("Unexpected stack trace: " +
-					Arrays.toString(trace), success);
 			}
 
-			//
-			// 3) Trigger sample for the blocking task with max depth
-			//
-			int maxDepth = 2;
-
-			CompletableFuture<StackTraceSampleResponse> successfulSampleFutureWithMaxDepth =
-				tmGateway.requestStackTraceSample(eid, sampleId3, numSamples, Time.milliseconds(100L), maxDepth, timeout);
-
-			StackTraceSampleResponse responseWithMaxDepth = successfulSampleFutureWithMaxDepth.get();
-
-			assertEquals(sampleId3, responseWithMaxDepth.getSampleId());
-			assertEquals(eid, responseWithMaxDepth.getExecutionAttemptID());
-
-			List<StackTraceElement[]> tracesWithMaxDepth = responseWithMaxDepth.getSamples();
-
-			assertEquals("Number of samples", numSamples, tracesWithMaxDepth.size());
-
-			for (StackTraceElement[] trace : tracesWithMaxDepth) {
-				assertEquals("Max depth", maxDepth, trace.length);
-			}
+			assertEquals("Task was not back pressured in given time.", 1.0, backPressureRatio, 0.0);
 
-			//
-			// 4) Trigger sample for the blocking task, but cancel it during sampling
-			//
-			int sleepTime = 100;
-			numSamples = 100;
+			// 3) trigger request for the blocking task, but cancel it before request finishes.
+			final int sleepTime = 1000;
 
-			CompletableFuture<StackTraceSampleResponse> canceldSampleFuture =
-				tmGateway.requestStackTraceSample(eid, sampleId4, numSamples, Time.milliseconds(10L), maxDepth, timeout);
+			CompletableFuture<TaskBackPressureResponse> canceledRequestFuture =
+				tmGateway.requestTaskBackPressure(executionAttemptID, requestId, timeout);
 
 			Thread.sleep(sleepTime);
```

Comment:
Why sleep here or how can be know how long to sleep?

Reply:
The sleep tries to ensure that the sample has started before canceling. I guess 1s is enough.

Reply:
Maybe we can ensure the sample has started by checking the stack trace instead of waiting, what do you think?

Reply:
Checking the stack trace seems not an elegant way unless necessary. The proper way should inject some custom implementations which can provide the signal when the condition is satisfied to trigger other actions. Considering it is not very easy to do that and this `sleep` mechanism is not brought by this PR, then we can keep it as now.